akka-stream/src/main/scala/akka/stream/impl/UnfoldResourceSourceAsync.scala

```diff
@@ -4,138 +4,140 @@
-
-    override def onPull(): Unit =
-      state match {
-        case Some(resource) =>
-          try {
-            val future = readData(resource)
-            future.value match {
-              case Some(value) => handle(value)
-              case None        => future.onComplete(readCallback)(parasitic)
-            }
-          } catch errorHandler
-        case None =>
-        // we got a pull but there is no open resource, we are either
-        // currently creating/restarting then the read will be triggered when creating the
-        // resource completes, or shutting down and then the pull does not matter anyway
+      private val readCallback = getAsyncCallback[Try[Option[T]]](handle).invoke _
+
+      private def handle(result: Try[Option[T]]): Unit = result match {
+        case Success(data) =>
+          data match {
+            case Some(d) => push(out, d)
+            case None    =>
+              // end of resource reached, lets close it
+              maybeResource match {
+                case OptionVal.Some(resource) =>
+                  close(resource).onComplete(getAsyncCallback[Try[Done]] {
+                    case Success(Done) => completeStage()
+                    case Failure(ex)   => failStage(ex)
+                  }.invoke)(parasitic)
+                  maybeResource = OptionVal.none
+
+                case _ =>
+                  // cannot happen, but for good measure
+                  throw new IllegalStateException("Reached end of data but there is no open resource")
+              }
+          }
+        case Failure(t) => errorHandler(t)
       }
 
-    override def postStop(): Unit = {
-      state.foreach(r => close(r))
-    }
+      override def preStart(): Unit = createResource()
+
+      override def onPull(): Unit =
+        maybeResource match {
+          case OptionVal.Some(resource) =>
+            try {
+              val future = readData(resource)
+              future.value match {
+                case Some(value) => handle(value)
+                case None        => future.onComplete(readCallback)(parasitic)
+              }
+            } catch errorHandler
+          case _ =>
+          // we got a pull but there is no open resource, we are either
+          // currently creating/restarting then the read will be triggered when creating the
+          // resource completes, or shutting down and then the pull does not matter anyway
+        }
 
-    private def restartResource(): Unit = {
-      state match {
-        case Some(resource) =>
-          // wait for the resource to close before restarting
-          close(resource).onComplete(getAsyncCallback[Try[Done]] {
-            case Success(Done) =>
-              createResource()
-            case Failure(ex) => failStage(ex)
-          }.invoke)(parasitic)
-          state = None
-        case None =>
-          createResource()
+      override def postStop(): Unit = {
+        maybeResource.foreach(r => close(r))
```

Comment:
I'm not 100% sure but I think the reason we didn't add methods like `map` to OptionVal was that it will box the value and defeat the purpose of 0 allocations when using OptionVal. Probably doesn't matter in such postStop but might be better to not have `foreach` so that someone is using it without knowing later.

Reply:
I was remember only be used as type parameter and added to a collection will box?

Reply:
I have dropped that commit.